contracts/voting/dao-voting-native-staked/src/contract.rs

```diff
@@ -64,29 +79,134 @@ pub fn instantiate(
     let config = Config {
         owner,
         manager,
-        denom: msg.denom,
         unstaking_duration: msg.unstaking_duration,
     };
 
     CONFIG.save(deps.storage, &config)?;
     DAO.save(deps.storage, &info.sender)?;
 
-    Ok(Response::new()
-        .add_attribute("action", "instantiate")
-        .add_attribute(
-            "owner",
-            config
-                .owner
-                .map(|a| a.to_string())
-                .unwrap_or_else(|| "None".to_string()),
-        )
-        .add_attribute(
-            "manager",
-            config
-                .manager
-                .map(|a| a.to_string())
-                .unwrap_or_else(|| "None".to_string()),
-        ))
+    if let Some(active_threshold) = msg.active_threshold.as_ref() {
+        if let ActiveThreshold::Percentage { percent } = active_threshold {
+            if *percent > Decimal::percent(100) || *percent <= Decimal::percent(0) {
+                return Err(ContractError::InvalidActivePercentage {});
+            }
+        }
+        ACTIVE_THRESHOLD.save(deps.storage, active_threshold)?;
+    }
+
+    match msg.token_info {
+        TokenInfo::Existing { denom } => {
+            if let Some(ActiveThreshold::AbsoluteCount { count }) = msg.active_threshold {
+                assert_valid_absolute_count_threshold(deps.as_ref(), &denom, count)?;
+            }
+
+            DENOM.save(deps.storage, &denom)?;
+
+            Ok(Response::new()
+                .add_attribute("action", "instantiate")
+                .add_attribute("token", "existing_token")
+                .add_attribute("token_denom", denom)
+                .add_attribute(
+                    "owner",
+                    config
+                        .owner
+                        .map(|a| a.to_string())
+                        .unwrap_or_else(|| "None".to_string()),
+                )
+                .add_attribute(
+                    "manager",
+                    config
+                        .manager
+                        .map(|a| a.to_string())
+                        .unwrap_or_else(|| "None".to_string()),
+                ))
+        }
+        TokenInfo::New {
+            tf_core_code_id,
+            info: denom_info,
+            initial_dao_balance,
+        } => {
+            let mut initial_balances = denom_info.initial_balances.unwrap_or_default();
+
+            let initial_supply = initial_balances
+                .iter()
+                .fold(Uint128::zero(), |p, n| p + n.amount);
```

Comment:
would be nice to add some comments or more descriptive closure parameter names here